src/Services/Device.vala

```diff
@@ -0,0 +1,342 @@
+        luminosity = upower_device.luminosity;
+        percentage = upower_device.percentage;
+        temperature = upower_device.temperature;
+        voltage = upower_device.voltage;
+        time_to_empty = upower_device.time_to_empty;
+        time_to_full = upower_device.time_to_full;
+        model = upower_device.model;
+        native_path = upower_device.native_path;
+        serial = upower_device.serial;
+        vendor = upower_device.vendor;
+        device_type = determine_device_type ();
+        state = (State) upower_device.state;
+        technology = (Technology) upower_device.technology;
+        update_time = upower_device.update_time;
+
+        is_charging = state == State.FULLY_CHARGED || state == State.CHARGING;
+        is_a_battery = device_type != Type.UNKNOWN && device_type != Type.LINE_POWER;
+
+        properties_updated ();
+    }
+
+    public bool is_present () {
+        bool present = false;
+        if (upower.on_battery || upower_device.is_present) {
+            present = true;
+        }
+
+        return present;
+    }
+
+    public string get_info () {
+        var percent = (int)Math.round (percentage);
+        if (percent <= 0) {
+            return _("Calculating…");
+        }
+
+        if (percent == 100 && is_charging) {
+            return _("Fully charged");
+        }
+
+        var info = _("%i%% charged").printf (percent);
+
+        return info;
+    }
+
+    public string get_icon_name_for_battery () {
+        if (!is_a_battery) {
+            return "preferences-system-power-symbolic";
+        }
+        if (percentage == 100 && is_charging) {
+            return "battery-full-charged";
+        }
+        var battery_icon = get_battery_icon ();
+        if (is_charging) {
+            return battery_icon + "-charging";
+        } else {
+            return battery_icon;
+        }
+    }
+
+    public string get_battery_icon () {
+        if (percentage <= 0) {
+            return "battery-good";
+        }
+
+        if (percentage < 10 && (time_to_empty == 0 || time_to_empty < 30 * 60)) {
+            return "battery-empty";
+        }
+
+        if (percentage < 30) {
+            return "battery-caution";
+        }
+
+        if (percentage < 60) {
+            return "battery-low";
+        }
+
+        if (percentage < 80) {
+            return "battery-good";
+        }
+
+        return "battery-full";
+    }
+
+    public string get_current_charge () {
+        return format_capacity (percentage, "%");
+    }
+
+    public string get_max_capacity () {
+        return format_capacity (capacity, "%");
+    }
+
+    public string get_design_energy () {
+        return format_capacity (energy_full_design, " Wh");
+    }
+
+    private static string format_capacity (double value, string unit) {
+        var value_int = (int)Math.round (value);
+        if (value_int == 0) {
+            return _("Unknown");
+        }
+
+        return value_int.to_string () + unit;
```

Comment:
Maybe we may not want to assert a unit comes after a value; IIRC "15%" should be *translated* into "%15" in Turkish although it's not RTL language: https://l10n.elementary.io/translate/wingpanel/wingpanel-indicator-power/tr/?checksum=91b9573aa836f687&sort_by=-priority,position

So we should make current charge, max capacity, and design energy translatable using gettext.

Reply:
Done !